include/range/v3/view/shared.hpp

```diff
@@ -0,0 +1,148 @@
+
+            // use the stored range's begin and end
+            iterator_t<Rng> begin() const
+            {
+                return ranges::begin(*rng_ptr_);
+            }
+            sentinel_t<Rng> end() const
+            {
+                return ranges::end(*rng_ptr_);
+            }
+
+            // use the const-most size() function provided by the range
+            CONCEPT_REQUIRES(SizedRange<const Rng>())
+            range_size_type_t<Rng> size() const
+            {
+                return ranges::size(*rng_ptr_);
+            }
+            CONCEPT_REQUIRES(SizedRange<Rng>() && !SizedRange<const Rng>())
+            range_size_type_t<Rng> size()
+            {
+                return ranges::size(*rng_ptr_);
+            }
+
+            // shared storage access
+            const std::shared_ptr<Rng> & get_shared() const
+            {
+                return rng_ptr_;
+            }
+        };
+
+        /// \relates all
+        /// \addtogroup group-views
+        /// @{
+        namespace view
+        {
+            struct shared_fn : pipeable<shared_fn>
+            {
+            public:
+                template<typename Rng,
+                    CONCEPT_REQUIRES_(Range<Rng>())>
+                shared_view<Rng> operator()(std::shared_ptr<Rng> pt) const
+                {
+                    return {std::move(pt)};
+                }
+
+#ifndef RANGES_DOXYGEN_INVOKED
+                template<typename Rng,
+                    CONCEPT_REQUIRES_(!Range<Rng>())>
+                void operator()(std::shared_ptr<Rng>) const
+                {
+                    CONCEPT_ASSERT_MSG(Range<Rng>(),
+                        "The object on which view::shared operates must be a "
+                        "model of the Range concept.");
+                }
+#endif
+
+                template<typename Rng,
+                    CONCEPT_REQUIRES_(Range<Rng>()
+                                      && !View<Rng>()
+                                      && !std::is_reference<Rng>::value)>
+                shared_view<Rng> operator()(Rng && t) const
+                {
+                    return shared_view<Rng>{std::move(t)};
+                }
+
+#ifndef RANGES_DOXYGEN_INVOKED
+                template<typename Rng,
+                    CONCEPT_REQUIRES_(!Range<Rng>()
+                                      || View<Rng>()
+                                      || std::is_reference<Rng>::value)>
+                void operator()(Rng &&) const
+                {
+                    CONCEPT_ASSERT_MSG(Range<Rng>(),
+                        "The object on which view::shared operates must be a "
+                        "model of the Range concept.");
+                    CONCEPT_ASSERT_MSG(!View<Rng>(),
+                        "view::shared cannot be constructed from a view."
+                        " Please copy the original view instead.");
+                    CONCEPT_ASSERT_MSG(std::is_rvalue_reference<Rng&&>::value,
```

Comment:
Better to mirror the constraint here with `CONCEPT_ASSERT_MSG(!std::is_reference<Rng>::value,...`.